File: app/src/main/java/com/soundcloud/android/image/ImageOperations.java
package com.soundcloud.android.image;

import static com.soundcloud.android.image.ImageOperations.DisplayType.CIRCULAR;

import com.nostra13.universalimageloader.core.DisplayImageOptions;
import com.nostra13.universalimageloader.core.ImageLoader;
import com.nostra13.universalimageloader.core.assist.ImageSize;
import com.nostra13.universalimageloader.core.assist.ViewScaleType;
import com.nostra13.universalimageloader.core.imageaware.ImageAware;
import com.nostra13.universalimageloader.core.imageaware.ImageViewAware;
import com.nostra13.universalimageloader.core.imageaware.NonViewAware;
import com.nostra13.universalimageloader.core.listener.ImageLoadingListener;
import com.nostra13.universalimageloader.core.listener.PauseOnScrollListener;
import com.nostra13.universalimageloader.utils.MemoryCacheUtils;
import com.soundcloud.android.R;
import com.soundcloud.android.model.Urn;
import com.soundcloud.android.utils.DeviceHelper;
import com.soundcloud.android.utils.images.ImageUtils;
import com.soundcloud.java.optional.Optional;
import rx.Observable;
import rx.Observable.OnSubscribe;
import rx.Scheduler;
import rx.Subscriber;
import rx.functions.Func1;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.TransitionDrawable;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.graphics.Palette;
import android.view.View;
import android.widget.AbsListView;
import android.widget.ImageView;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

@Singleton
public class ImageOperations {

    public enum DisplayType {
        DEFAULT, CIRCULAR
    }

    private static final Func1<Bitmap, Observable<Palette>> BITMAP_TO_PALETTE = bitmap -> Observable.create(
            (OnSubscribe<Palette>) subscriber -> Palette.from(bitmap)
                                                        .generate(palette -> {
                                                            if (!subscriber.isUnsubscribed()) {
                                                                subscriber.onNext(
                                                                        palette);
                                                                subscriber.onCompleted();
                                                            }
                                                        }));
    private final ImageLoader imageLoader;
    private final ImageUrlBuilder imageUrlBuilder;
    private final Set<String> notFoundUris = new HashSet<>();
    private final FallbackBitmapLoadingAdapter.Factory adapterFactory;
    private final BitmapLoadingAdapter.Factory bitmapAdapterFactory;
    private final DeviceHelper deviceHelper;
    private final PlaceholderGenerator placeholderGenerator;
    private final CircularPlaceholderGenerator circularPlaceholderGenerator;
    private final ImageCache imageCache;
    private final FallbackImageListener notFoundListener = new FallbackImageListener(notFoundUris);
    private final ImageProcessor imageProcessor;

    @Inject
    public ImageOperations(
            ImageLoader imageLoader,
            ImageUrlBuilder imageUrlBuilder,
            FallbackBitmapLoadingAdapter.Factory adapterFactory,
            BitmapLoadingAdapter.Factory bitmapAdapterFactory,
            ImageProcessor imageProcessor,
            DeviceHelper deviceHelper,
            PlaceholderGenerator placeholderGenerator,
            CircularPlaceholderGenerator circularPlaceholderGenerator,
            ImageCache imageCache) {
        this.imageUrlBuilder = imageUrlBuilder;
        this.adapterFactory = adapterFactory;
        this.bitmapAdapterFactory = bitmapAdapterFactory;
        this.imageProcessor = imageProcessor;
        this.deviceHelper = deviceHelper;
        this.placeholderGenerator = placeholderGenerator;
        this.circularPlaceholderGenerator = circularPlaceholderGenerator;
        this.imageCache = imageCache;

        this.imageLoader = imageLoader;
        this.imageLoader.init(imageCache.getImageLoaderConfiguration());
    }

    public void clearDiskCache() {
        imageLoader.clearDiskCache();
    }

    public void displayInAdapterView(Urn urn, Optional<String> imageUrlTemplate, ApiImageSize apiImageSize, ImageView imageView, DisplayType displayType) {
        displayInAdapterView(apiImageSize,
                             imageView,
                             getImageUrl(imageUrlTemplate, urn, apiImageSize),
                             notFoundListener,
                             Optional.absent(),
                             displayType);
    }

    public Observable<Bitmap> displayInAdapterView(final ImageResource imageResource,
                                                   final ApiImageSize apiImageSize,
                                                   final ImageView imageView,
                                                   final Optional<Drawable> fallbackDrawable,
                                                   DisplayType displayType) {
        return Observable.create(subscriber -> displayInAdapterView(apiImageSize,
                                                                    imageView,
                                                                    getImageUrl(imageResource.getImageUrlTemplate(), imageResource.getUrn(), apiImageSize),
                                                                    buildFallbackImageListener(fromBitmapSubscriber(subscriber)),
                                                                    fallbackDrawable,
                                                                    displayType));
    }

    private void displayInAdapterView(ApiImageSize apiImageSize,
                                      ImageView imageView,
                                      String imageUrl,
                                      FallbackImageListener imageListener,
                                      Optional<Drawable> placeholderDrawable,
                                      DisplayType displayType) {
        final ImageViewAware imageAware = new ImageViewAware(imageView, false);
        final Drawable drawable = placeholderDrawable.isPresent() ?
                                  placeholderDrawable.get() :
                                  getPlaceholderDrawable(imageUrl, imageAware.getWidth(), imageAware.getHeight(), displayType);

        DisplayImageOptions options = ImageOptionsFactory.adapterView(drawable, apiImageSize, deviceHelper);
        if (displayType.equals(DisplayType.CIRCULAR) && placeholderDrawable.isPresent()) {
            options = ImageOptionsFactory.adapterViewCircular(placeholderDrawable.get(), apiImageSize, deviceHelper);
        }
        imageLoader.displayImage(imageUrl, imageAware, options, imageListener);
    }

    public Observable<Palette> displayCircularInAdapterViewAndGeneratePalette(final ImageResource imageResource,
                                                                              final ApiImageSize apiImageSize,
                                                                              final ImageView imageView) {
        return displayInAdapterView(imageResource, apiImageSize, imageView, Optional.absent(), DisplayType.CIRCULAR).flatMap(BITMAP_TO_PALETTE);
    }

    public void displayDefaultPlaceholder(ImageView imageView) {
        displayWithPlaceholder(imageView, null, Optional.absent());
    }

    public void displayWithPlaceholder(Urn urn, Optional<String> imageUrlTemplate, ApiImageSize apiImageSize, ImageView imageView) {
        displayWithPlaceholder(imageView, getImageUrl(imageUrlTemplate, urn, apiImageSize), Optional.absent());
    }

    private void displayWithPlaceholder(ImageView imageView, @Nullable String imageUrl, Optional<ImageLoadingListener> imageListener) {
        final ImageViewAware imageAware = new ImageViewAware(imageView, false);
        imageLoader.displayImage(
                imageUrl,
                imageAware,
                ImageOptionsFactory.placeholder(getPlaceholderDrawable(imageUrl, imageAware.getWidth(), imageAware.getHeight(), DisplayType.DEFAULT)),
                imageListener.isPresent() ? imageListener.get() : notFoundListener);
    }

    public Observable<Bitmap> displayWithPlaceholderObservable(ImageResource imageResource, ApiImageSize apiImageSize, ImageView imageView) {
        return Observable.create(subscriber -> displayWithPlaceholder(imageView,
                                                                      getImageUrl(imageResource.getImageUrlTemplate(), imageResource.getUrn(), apiImageSize),
                                                                      Optional.of(buildFallbackImageListener(bitmapAdapterFactory.create(subscriber)))));
    }


    public void displayCircular(String imageUrl, ImageView imageView) {
        imageLoader.displayImage(imageUrl, new ImageViewAware(imageView, false),
                                 ImageOptionsFactory.placeholderCircular(imageView.getResources()
                                                                                  .getDrawable(R.drawable.circular_placeholder)));
    }

    public void displayCircularWithPlaceholder(Urn urn, Optional<String> imageUrlTemplate,
                                               ApiImageSize apiImageSize,
                                               ImageView imageView) {
        String imageUrl = getImageUrl(imageUrlTemplate, urn, apiImageSize);
        final ImageViewAware imageAware = new ImageViewAware(imageView, false);
        final DisplayImageOptions options = ImageOptionsFactory.placeholderCircular(
                imageCache.getPlaceholderDrawable(imageUrl, imageAware.getWidth(), imageAware.getHeight(), circularPlaceholderGenerator));
        imageLoader.displayImage(
                imageUrl,
                imageAware,
                options,
                notFoundListener);
    }

    public void displayInPlayer(ImageResource imageResource,
                                ApiImageSize apiImageSize,
                                ImageView imageView,
                                Bitmap placeholder,
                                boolean isHighPriority) {
        final ImageViewAware imageAware = new ImageViewAware(imageView, false);
        final Drawable placeholderDrawable = placeholder != null ? new BitmapDrawable(placeholder) :
                                             getPlaceholderDrawable(imageResource.getUrn().toString(), imageAware.getWidth(), imageAware.getHeight(), DisplayType.DEFAULT);

        imageLoader.displayImage(
                getImageUrl(imageResource.getImageUrlTemplate(), imageResource.getUrn(), apiImageSize),
                imageAware,
                ImageOptionsFactory.player(placeholderDrawable, isHighPriority),
                notFoundListener);
    }

    public void displayAdImage(Urn urn, String imageUri, ImageView imageView, final ImageListener listener) {
        displayAdImage(urn, imageUri, imageView, new ImageListenerUILAdapter(listener));
    }

    public void displayLeaveBehind(Uri uri, ImageView imageView, ImageListener imageListener) {
        final ImageViewAware imageAware = new ImageViewAware(imageView, false);
        imageLoader.displayImage(
                uri.toString(),
                imageAware,
                ImageOptionsFactory.adImage(),
                new ImageListenerUILAdapter(imageListener));
    }

    public void displayInFullDialogView(ImageResource imageResource,
                                        ApiImageSize apiImageSize,
                                        ImageView imageView,
                                        ImageListener imageListener) {
        imageLoader.displayImage(
                getImageUrl(imageResource.getImageUrlTemplate(), imageResource.getUrn(), apiImageSize),
                new ImageViewAware(imageView, false),
                ImageOptionsFactory.fullImageDialog(),
                new ImageListenerUILAdapter(imageListener));
    }

    public Observable<Bitmap> bitmap(final Uri uri) {
        return Observable.create(subscriber -> {
            // We pass NonViewAware to circumvent ImageLoader cancelling requests (https://github.com/nostra13/Android-Universal-Image-Loader/issues/681)
            imageLoader.displayImage(uri.toString(),
                                     new NonViewAware(new ImageSize(0, 0), ViewScaleType.CROP),
                                     ImageOptionsFactory.adImage(),
                                     new ImageListenerUILAdapter(bitmapAdapterFactory.create(subscriber)));
        });
    }

    public Observable<Bitmap> artwork(final ImageResource imageResource, final ApiImageSize apiImageSize) {
        return Observable.create(subscriber -> {
            final Bitmap fallback = createFallbackBitmap(imageResource.getUrn(), apiImageSize);
            imageLoader.loadImage(
                    getImageUrl(imageResource.getImageUrlTemplate(), imageResource.getUrn(), apiImageSize),
                    new ImageListenerUILAdapter(adapterFactory.create(subscriber, fallback)));
        });
    }

    public Observable<Bitmap> artwork(final ImageResource imageResource,
                                      final ApiImageSize apiImageSize,
                                      final int targetWidth,
                                      final int targetHeight) {
        return Observable.create(subscriber -> {
            final GradientDrawable fallbackDrawable = generateDrawable(imageResource);
            final Bitmap fallback = ImageUtils.toBitmap(fallbackDrawable, targetWidth, targetHeight);
            load(imageResource,
                 apiImageSize,
                 targetWidth,
                 targetHeight,
                 adapterFactory.create(subscriber, fallback));
        });
    }

    public void precacheArtwork(ImageResource imageResource, ApiImageSize apiImageSize) {
        String url = getImageUrl(imageResource.getImageUrlTemplate(), imageResource.getUrn(), apiImageSize);
        imageLoader.loadImage(url, ImageOptionsFactory.prefetch(), null);
    }

    @Nullable
    public Bitmap getCachedListItemBitmap(Resources resources, ImageResource imageResource) {
        return getCachedBitmap(imageResource, ApiImageSize.getListItemImageSize(resources),
                               resources.getDimensionPixelSize(R.dimen.list_item_image_dimension),
                               resources.getDimensionPixelSize(R.dimen.list_item_image_dimension));
    }

    @Nullable
    public Bitmap getCachedBitmap(ImageResource imageResource,
                                  ApiImageSize apiImageSize,
                                  int targetWidth,
                                  int targetHeight) {
        final String imageUrl = imageUrlBuilder.buildUrl(imageResource.getImageUrlTemplate(), imageResource.getUrn(), apiImageSize);
        final String key = MemoryCacheUtils.generateKey(imageUrl, new ImageSize(targetWidth, targetHeight));
        return imageLoader.getMemoryCache().get(key);
    }

    public AbsListView.OnScrollListener createScrollPauseListener(boolean pauseOnScroll, boolean pauseOnFling,
                                                                  AbsListView.OnScrollListener customListener) {
        return new PauseOnScrollListener(imageLoader, pauseOnScroll, pauseOnFling, customListener);
    }

    @Nullable
    public Bitmap decodeResource(Resources resources, int resId) {
        try {
            return BitmapFactory.decodeResource(resources, resId);
        } catch (OutOfMemoryError e) {
            e.printStackTrace();
            return null;
        }
    }

    private DefaultImageListener fromBitmapSubscriber(final Subscriber<? super Bitmap> subscriber) {
        return new DefaultImageListener() {
            @Override
            public void onLoadingFailed(String imageUri,
                                        View view,
                                        @Nullable Throwable cause) {
                if (!subscriber.isUnsubscribed()) {
                    subscriber.onError(cause == null ? new IOException("Failed to load bitmap for Unknown reason") : cause);
                }
            }

            @Override
            public void onLoadingComplete(String imageUri,
                                          View view,
                                          Bitmap loadedImage) {
                if (subscriber.isUnsubscribed()) {
                    return;
                }

                if (imageUri == null && loadedImage == null) {
                    subscriber.onError(new IOException(
                            "Image loading failed."));
                    return;
                }
                subscriber.onNext(loadedImage);
                subscriber.onCompleted();
            }
        };
    }

    private void displayAdImage(Urn urn, String imageUri, ImageView imageView, ImageLoadingListener listener) {
        final ImageViewAware imageAware = new ImageViewAware(imageView, false);
        final Drawable drawable = getPlaceholderDrawable(urn.toString(), imageAware.getWidth(), imageAware.getHeight(), DisplayType.DEFAULT);
        final DisplayImageOptions options = ImageOptionsFactory.streamAdImage(drawable, deviceHelper);

        imageLoader.displayImage(imageUri, imageAware, options, listener);
    }

    private void load(ImageResource imageResource,
                      ApiImageSize apiImageSize,
                      int targetWidth,
                      int targetHeight,
                      ImageListener imageListener) {
        ImageSize targetSize = new ImageSize(targetWidth, targetHeight);
        ImageAware imageAware = new NonViewAware(targetSize, ViewScaleType.CROP);
        imageLoader.displayImage(getImageUrl(imageResource.getImageUrlTemplate(), imageResource.getUrn(), apiImageSize),
                                 imageAware,
                                 new ImageListenerUILAdapter(imageListener));
    }

    private Func1<Bitmap, Bitmap> blurBitmap(final Optional<Float> blurRadius) {
        return bitmap -> imageProcessor.blurBitmap(bitmap, blurRadius);
    }


    private Observable<Bitmap> blurBitmap(final Bitmap original, final Optional<Float> blurRadius) {
        return Observable.create(subscriber -> {
            subscriber.onNext(imageProcessor.blurBitmap(original, blurRadius));
            subscriber.onCompleted();
        });
    }

    @Nullable
    public String getImageUrl(Optional<String> imageUrlTemplate, Urn urn, ApiImageSize apiImageSize) {
        final String imageUrl = imageUrlBuilder.buildUrl(imageUrlTemplate, urn, apiImageSize);
        return notFoundUris.contains(imageUrl) ? null : imageUrl;
    }

    @NonNull
    private FallbackImageListener buildFallbackImageListener(ImageListener imageListener) {
        return new FallbackImageListener(imageListener, notFoundUris);
    }

    public Observable<Bitmap> blurredArtwork(final Resources resources,
                                             final ImageResource imageResource,
                                             Optional<Float> blurRadius,
                                             Scheduler scheduleOn, Scheduler observeOn) {
        final Bitmap cachedBlurImage = imageCache.getBlurredImage(imageResource.getUrn());
        if (cachedBlurImage != null) {
            return Observable.just(cachedBlurImage);
        } else {
            final Bitmap cached = getCachedListItemBitmap(resources, imageResource);
            if (cached == null) {
                return artwork(imageResource, ApiImageSize.getListItemImageSize(resources))
                        .map(blurBitmap(blurRadius))
                        .subscribeOn(scheduleOn)
                        .observeOn(observeOn)
                        .doOnNext(imageCache.cacheBlurredBitmap(imageResource.getUrn()));
            } else {
                return blurBitmap(cached, blurRadius)
                        .subscribeOn(scheduleOn)
                        .observeOn(observeOn)
                        .doOnNext(imageCache.cacheBlurredBitmap(imageResource.getUrn()));
            }
        }
    }

    GradientDrawable generateDrawable(ImageResource imageResource) {
        return placeholderGenerator.generateDrawable(imageResource.getUrn().toString());
    }


    @Nullable
    TransitionDrawable getPlaceholderDrawable(@Nullable final String imageUrl, int width, int height, ImageOperations.DisplayType displayType) {
        PlaceholderGenerator placeholderGenerator = displayType.equals(CIRCULAR) ? this.circularPlaceholderGenerator : this.placeholderGenerator;
        return imageCache.getPlaceholderDrawable(imageUrl,
                                                 width,
                                                 height,
                                                 placeholderGenerator);
    }

    private Bitmap createFallbackBitmap(Urn resourceUrn, ApiImageSize apiImageSize) {
        // This bitmap is only used by the current track for the components that can't use
        // drawables (i.e. the notification and the remote client for the lock screen)
        //
        // Unless we refactor the cache to store a /GradientDrawable/ and not a /TransitionDrawable/
        // we don't have a cache for this guy.
        //
        // Also, we don't cache bitmap in the /ImageOperations/ since it does not worth it. A cache
        // may have a impact on the memory usage and without the performance seems pretty good, though.
        final GradientDrawable fallbackDrawable = placeholderGenerator.generateDrawable(resourceUrn.toString());
        return ImageUtils.toBitmap(fallbackDrawable, apiImageSize.width, apiImageSize.height);
    }

    void resume() {
        imageLoader.resume();
    }

    void pause() {
        imageLoader.pause();
    }
}

File: app/src/main/java/com/soundcloud/android/image/ImageOperations.kt
package com.soundcloud.android.image

import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import android.support.v7.graphics.Palette
import android.widget.AbsListView
import android.widget.ImageView
import com.soundcloud.android.R
import com.soundcloud.android.model.Urn
import com.soundcloud.android.rx.observers.LambdaObserver
import com.soundcloud.android.utils.OpenForTesting
import com.soundcloud.android.utils.images.ImageUtils
import com.soundcloud.java.optional.Optional
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.Single
import io.reactivex.SingleEmitter
import io.reactivex.disposables.CompositeDisposable
import java.io.IOException
import java.util.HashSet
import javax.inject.Inject
import javax.inject.Singleton

@OpenForTesting
@Singleton
class ImageOperations
@Inject
constructor(
        private val imageLoader: ImageLoader,
        private val imageUrlBuilder: ImageUrlBuilder,
        private val imageProcessor: ImageProcessor,
        private val placeholderGenerator: PlaceholderGenerator,
        private val imageCache: ImageCache) {
    private val notFoundUris = HashSet<String>()
    private val notFoundConsumer = FallbackImageConsumer(notFoundUris = notFoundUris)
    private val compositeDisposable = CompositeDisposable()

    fun clearDiskCache() {
        imageLoader.clearDiskCache()
    }

    fun displayInAdapterView(urn: Urn, imageUrlTemplate: Optional<String> = Optional.absent(), apiImageSize: ApiImageSize, imageView: ImageView, circular: Boolean) {
        compositeDisposable.add(imageLoader.displayImage(toImageUrl(urn = urn, imageUrlTemplate = imageUrlTemplate.orNull(), apiImageSize = apiImageSize),
                                                         imageView = imageView,
                                                         circular = circular,
                                                         apiImageSize = apiImageSize)
                                        .subscribeWith(LambdaObserver.onNext { notFoundConsumer }))
    }

    fun displayInAdapterViewSingle(urn: Urn,
                                   imageUrlTemplate: Optional<String> = Optional.absent(),
                                   apiImageSize: ApiImageSize,
                                   imageView: ImageView,
                                   fallbackDrawable: Optional<Drawable> = Optional.absent(),
                                   circular: Boolean): Single<Bitmap> {
        return imageLoader.displayImage(toImageUrl(urn, imageUrlTemplate.orNull(), apiImageSize),
                                        imageView,
                                        circular = circular,
                                        placeholderDrawable = fallbackDrawable.orNull(),
                                        apiImageSize = apiImageSize)
                .compose(this::toBitmap)
                .firstOrError()
    }

    fun displayCircularInAdapterViewAndGeneratePalette(urn: Urn,
                                                       imageUrlTemplate: Optional<String>,
                                                       apiImageSize: ApiImageSize,
                                                       imageView: ImageView): Single<Palette> {
        return displayInAdapterViewSingle(urn, imageUrlTemplate, apiImageSize, imageView, circular = true)
                .flatMap { bitmap: Bitmap ->
                    Single.create { subscriber: SingleEmitter<Palette> ->
                        Palette.from(bitmap).generate { palette -> subscriber.onSuccess(palette) }
                    }
                }
    }

    fun displayDefaultPlaceholder(imageView: ImageView) {
        compositeDisposable.add(displayWithPlaceholder(imageView)
                                        .subscribeWith(LambdaObserver.onNext(notFoundConsumer)))
    }

    fun displayWithPlaceholder(urn: Urn, imageUrlTemplate: Optional<String>, apiImageSize: ApiImageSize, imageView: ImageView) {
        compositeDisposable.add(displayWithPlaceholder(imageView, toImageUrl(urn, imageUrlTemplate.orNull(), apiImageSize), apiImageSize = apiImageSize)
                                        .subscribeWith(LambdaObserver.onNext(notFoundConsumer)))
    }

    private fun displayWithPlaceholder(imageView: ImageView, imageUrl: String? = null, apiImageSize: ApiImageSize = ApiImageSize.Unknown): Observable<LoadingState> {
        return imageLoader.displayImage(imageUrl,
                                        imageView,
                                        displayType = DisplayType.PLACEHOLDER,
                                        apiImageSize = apiImageSize)
    }

    fun displayWithPlaceholderObservable(urn: Urn, imageUrlTemplate: Optional<String> = Optional.absent(), apiImageSize: ApiImageSize, imageView: ImageView): Single<Bitmap> {
        return displayWithPlaceholder(imageView,
                                      toImageUrl(urn, imageUrlTemplate.orNull(), apiImageSize),
                                      apiImageSize = apiImageSize)
                .compose(this::toBitmap)
                .firstOrError()
    }

    private fun toBitmap(input: Observable<LoadingState>): Observable<Bitmap> {
        return input.doOnNext { notFoundConsumer }
                .doOnNext(this::throwErrorOnMissingBitmap)
                .ofType(LoadingState.Complete::class.java)
                .map { it.loadedImage }
    }

    fun displayCircular(imageUrl: String, imageView: ImageView) {
        compositeDisposable.add(imageLoader.displayImage(imageUrl, imageView, circular = true, apiImageSize = ApiImageSize.T120).subscribeWith(LambdaObserver.onNext(notFoundConsumer)))
    }

    fun displayCircularWithPlaceholder(urn: Urn,
                                       imageUrlTemplate: Optional<String>,
                                       apiImageSize: ApiImageSize,
                                       imageView: ImageView) {
        compositeDisposable.add(imageLoader.displayImage(toImageUrl(urn, imageUrlTemplate.orNull(), apiImageSize),
                                                         imageView,
                                                         circular = true,
                                                         apiImageSize = apiImageSize)
                                        .subscribeWith(LambdaObserver.onNext(notFoundConsumer)))
    }

    fun displayInPlayer(urn: Urn,
                        imageUrlTemplate: Optional<String> = Optional.absent(),
                        apiImageSize: ApiImageSize,
                        imageView: ImageView,
                        placeholder: Bitmap?,
                        isHighPriority: Boolean) {
        compositeDisposable.add(imageLoader.displayImage(toImageUrl(urn, imageUrlTemplate.orNull(), apiImageSize),
                                                         imageView,
                                                         placeholderDrawable = placeholder?.let { BitmapDrawable(imageView.resources, it) },
                                                         displayType = DisplayType.PLAYER,
                                                         apiImageSize = apiImageSize,
                                                         isHighPriority = isHighPriority)
                                        .subscribeWith(LambdaObserver.onNext(notFoundConsumer)))
    }

    fun displayLeaveBehind(uri: String, imageView: ImageView): Observable<LoadingState> {
        return imageLoader.displayImage(uri, imageView, displayType = DisplayType.AD)
    }

    fun displayInFullDialogView(urn: Urn,
                                imageUrlTemplate: Optional<String>,
                                apiImageSize: ApiImageSize,
                                imageView: ImageView): Observable<LoadingState> {
        return imageLoader.displayImage(toImageUrl(urn, imageUrlTemplate.orNull(), apiImageSize),
                                        imageView,
                                        displayType = DisplayType.FULL_IMAGE_DIALOG,
                                        apiImageSize = apiImageSize)
    }

    fun bitmap(uri: Uri, loadType: LoadType): Single<Bitmap> = imageLoader.loadImage(uri.toString(), loadType = loadType).compose(this::toBitmap).firstOrError()

    fun bitmap(urn: Urn, apiImageSize: ApiImageSize, loadType: LoadType): Maybe<Bitmap> {
        val imageUrl = toImageUrl(urn = urn, apiImageSize = apiImageSize)

        imageUrl?.let {
            return imageLoader.loadImage(it, loadType = loadType).compose(this::toBitmap).firstOrError().toMaybe()
        }

        return Maybe.empty()
    }

    fun artwork(urn: Urn, imageUrlTemplate: Optional<String>, apiImageSize: ApiImageSize): Maybe<Bitmap> {
        val imageUrl = toImageUrl(urn = urn, imageUrlTemplate = imageUrlTemplate.orNull(), apiImageSize = apiImageSize)

        imageUrl?.let {
            return imageLoader.loadImage(it).compose { toFallback(it, { createFallbackBitmap(urn, apiImageSize.width, apiImageSize.height) }) }.firstElement()
        }

        return Maybe.empty()
    }

    private fun toFallback(input: Observable<LoadingState>, createFallbackBitmap: () -> Bitmap): Observable<Bitmap> {
        return input.flatMapMaybe {
            when (it) {
                is LoadingState.Complete -> Maybe.just(it.loadedImage ?: createFallbackBitmap())
                is LoadingState.Start -> Maybe.empty()
                is LoadingState.Cancel -> Maybe.empty()
                is LoadingState.Fail -> Maybe.just(createFallbackBitmap())
            }
        }
    }

    fun artwork(urn: Urn,
                imageUrlTemplate: Optional<String>,
                apiImageSize: ApiImageSize,
                targetWidth: Int,
                targetHeight: Int): Single<Bitmap> {
        return load(urn, imageUrlTemplate.orNull(), apiImageSize, targetWidth, targetHeight).compose { toFallback(it, { createFallbackBitmap(urn, targetWidth, targetHeight) }) }.firstOrError()
    }

    fun precacheArtwork(urn: Urn,
                        imageUrlTemplate: Optional<String>,
                        apiImageSize: ApiImageSize) {
        toImageUrl(urn, imageUrlTemplate.orNull(), apiImageSize)?.let {
            imageLoader.loadImage(it, loadType = LoadType.PREFETCH)
        }?.let {
            compositeDisposable.add(it.subscribeWith(LambdaObserver.onNext(notFoundConsumer)))
        }
    }

    fun getCachedListItemBitmap(resources: Resources, urn: Urn, imageUrlTemplate: Optional<String>): Bitmap? {
        return getCachedBitmap(urn, imageUrlTemplate, ApiImageSize.getListItemImageSize(resources),
                               resources.getDimensionPixelSize(R.dimen.list_item_image_dimension),
                               resources.getDimensionPixelSize(R.dimen.list_item_image_dimension))
    }

    fun getCachedBitmap(urn: Urn,
                        imageUrlTemplate: Optional<String>,
                        apiImageSize: ApiImageSize,
                        targetWidth: Int,
                        targetHeight: Int): Bitmap? {

        val imageUrl = imageUrlBuilder.buildUrl(imageUrlTemplate.orNull(), urn, apiImageSize)
        return imageUrl?.let { imageLoader.getCachedBitmap(it, targetWidth, targetHeight) }
    }

    fun createScrollPauseListener(pauseOnScroll: Boolean, pauseOnFling: Boolean, customListener: AbsListView.OnScrollListener?): AbsListView.OnScrollListener? =
            imageLoader.createScrollPauseListener(pauseOnScroll, pauseOnFling, customListener)

    fun decodeResource(resources: Resources, resId: Int): Bitmap? {
        try {
            return BitmapFactory.decodeResource(resources, resId)
        } catch (e: OutOfMemoryError) {
            e.printStackTrace()
            return null
        }

    }

    private fun throwErrorOnMissingBitmap(loadingState: LoadingState) {
        when (loadingState) {
            is LoadingState.Fail -> throw BitmapLoadingAdapter.BitmapLoadingException(loadingState.cause)
            is LoadingState.Complete -> {
                if (loadingState.loadedImage == null) {
                    throw BitmapLoadingAdapter.BitmapLoadingException(IOException("Image loading failed."))
                }
            }
        }
    }

    fun displayAdImage(urn: Urn, imageUri: String, imageView: ImageView): Observable<LoadingState> {
        return imageLoader.displayImage(imageUri, imageView, displayType = DisplayType.STREAM_AD_IMAGE)
    }

    private fun load(urn: Urn,
                     imageUrlTemplate: String?,
                     apiImageSize: ApiImageSize,
                     targetWidth: Int,
                     targetHeight: Int): Observable<LoadingState> {
        val observable = toImageUrl(urn, imageUrlTemplate, apiImageSize)?.let {
            imageLoader.loadImage(it)
        }
        return observable ?: Observable.empty<LoadingState>()
    }

    private fun blurBitmap(original: Bitmap, blurRadius: Optional<Float>): Single<Bitmap> {
        return Single.create { subscriber ->
            subscriber.onSuccess(imageProcessor.blurBitmap(original, blurRadius))
        }
    }

    fun getImageUrl(urn: Urn, apiImageSize: ApiImageSize) = toImageUrl(urn = urn, apiImageSize = apiImageSize)

    private fun toImageUrl(urn: Urn, imageUrlTemplate: String? = null, apiImageSize: ApiImageSize): String? {
        val imageUrl = imageUrlBuilder.buildUrl(imageUrlTemplate, urn, apiImageSize)
        return if (notFoundUris.contains(imageUrl)) null else imageUrl
    }

    fun blurredArtwork(resources: Resources,
                       urn: Urn,
                       imageUrlTemplate: Optional<String>,
                       blurRadius: Optional<Float>,
                       scheduleOn: Scheduler, observeOn: Scheduler): Single<Bitmap> {
        val cachedBlurImage = imageCache.getBlurredImage(urn)
        if (cachedBlurImage != null) {
            return Single.just(cachedBlurImage)
        } else {
            val cached = getCachedListItemBitmap(resources, urn, imageUrlTemplate)
            return if (cached == null) {
                artwork(urn, imageUrlTemplate, ApiImageSize.getListItemImageSize(resources))
                        .map { imageProcessor.blurBitmap(it, blurRadius) }
                        .subscribeOn(scheduleOn)
                        .observeOn(observeOn)
                        .doOnSuccess(imageCache.cacheBlurredBitmap(urn))
                        .toSingle()
            } else {
                blurBitmap(cached, blurRadius)
                        .subscribeOn(scheduleOn)
                        .observeOn(observeOn)
                        .doOnSuccess(imageCache.cacheBlurredBitmap(urn))
            }
        }
    }

    private fun createFallbackBitmap(resourceUrn: Urn, width: Int, height: Int): Bitmap {
        // This bitmap is only used by the current track for the components that can't use
        // drawables (i.e. the notification and the remote client for the lock screen)
        //
        // Unless we refactor the cache to store a /GradientDrawable/ and not a /TransitionDrawable/
        // we don't have a cache for this guy.
        //
        // Also, we don't cache bitmap in the /ImageOperations/ since it does not worth it. A cache
        // may have a impact on the memory usage and without the performance seems pretty good, though.
        val fallbackDrawable = placeholderGenerator.generateDrawable(resourceUrn.toString())
        return ImageUtils.toBitmap(fallbackDrawable, width, height)
    }

    fun resume() {
        imageLoader.resume()
    }

    fun pause() {
        imageLoader.pause()
    }
}
